### ISSUE TITLE
Request through Kong periodically return '404 - no Route matched with those values'

### ISSUE THREAD

**Issue by dkelsey** [Kong/kong#5679]
Our issue seems very close to #5374 (Kong throws '404 - no Route matched with those values' because of corrupted Cache)

The differences are:
* We are on Kong `0.14.1`
* We are not making changes to routes yet we've observed that one route seems return `404` periodically.
* our `mem_cache_size` is set to `512m`
* There are other differences in some settings.

We cannot upgrade at this time due to budget.
I suspect we need to increase our `mem_cache_size`.

# Steps To Reproduce

We have a javascript app which makes CORS requests to an API - a Geocoder and a Router API.
The correct CORS headers are passed along with the request and the the CORS plugin is enabled and configured correctly.   This has been verified by making the same requests using curl.

Using the Javascript app, It is difficult to reproduce reliably: it works sometimes and not others. We have 3 people manually attempting to do this.

# Additional Details & Logs
* We are running 3 kong nodes and using cassandra as the data store.
* This is using dns round robin to load balance between kong nodes.
* The issue occurs on all nodes - off and on.
* The nodes are configured identically.

### `kong.conf`
_Note: settings that are commented out are defaults_
```
#cassandra_schema_consensus_timeout = 10000  
#cassandra_repl_factor = 1 
cassandra_data_centers = dc1:2,dc2:2
cassandra_local_datacenter = dc1
db_update_frequency = 5  
db_update_propagation = 15 
mem_cache_size = 512m 
ssl = on 
#cassandra_ssl_verify = off 
#cassandra_timeout = 5000 
anonymous_reports = off 
cassandra_consistency = ONE  
cassandra_lb_policy = DCAwareRoundRobin  
cassandra_repl_strategy = NetworkTopologyStrategy  
```

### `nginx-kong.conf`
```
lua_shared_dict kong                5m;
lua_shared_dict kong_db_cache       512m;
lua_shared_dict kong_db_cache_miss 12m;
lua_shared_dict kong_locks          8m;
lua_shared_dict kong_process_events 5m;
lua_shared_dict kong_cluster_events 5m;
lua_shared_dict kong_healthchecks   5m;
lua_shared_dict kong_rate_limiting_counters 12m;
lua_shared_dict kong_cassandra      5m;
lua_socket_log_errors off;
```

### Questions
* How can I estimate the optimal size of `mem_cache_size` based on configured apis, upstreams, routes, services and plugins?
* How can it get visibility into the cache behavior? 
* Why is a route that we have not touched being evicted or is not found?

**Comment by bungle** [comment]
Can you try to backport this:
https://github.com/Kong/kong/pull/5564

**Comment by dkelsey** [comment]
thanks.  looking at it now.

**Comment by bungle** [comment]
We have not backported it to 0.1x.x.

**Comment by dkelsey** [comment]
I'll take a look at the code.  What is described as the solution makes sense: thread clobbering shared variables.

**Comment by dkelsey** [comment]
I got my `0.14.1` environment setup in Vagrant.
When I looked at the `kong/db/strategies/cassandra/init.lua` I realized that `0.14.1` (the version we're using) is way, way behind.
Applying the describe solution is not trivial.

I learned where the error message we're seeing is being emitted from.  
I have an idea to add logging to determine the extent of the number of `'no Route and no API found with those values'` incidents.
`kong/runloop/handler.lua` line `484`

**Comment by bungle** [comment]
It should not be that hard:
1. Remove
https://github.com/Kong/kong/blob/85b17fdd57908829d3857631a0aa0e380f4f8db4/kong/db/strategies/cassandra/init.lua#L619

2. Add following to: https://github.com/Kong/kong/blob/85b17fdd57908829d3857631a0aa0e380f4f8db4/kong/db/strategies/cassandra/init.lua#L655
```
local opts = new_tab(0, 2)
```

And that's it.
```

**Comment by dkelsey** [comment]
Thanks I'll give it a try.

**Comment by ugolas** [comment]
@bungle Hi, will the fix at #5564 be added to 1.x at some point?

**Comment by guanlan** [comment]
> @bungle Hi, will the fix at #5564 be added to 1.x at some point?

It has been released with 2.0.2, and unlikely been backported to the older versions. Would like to help if you have any migration issue.

**Comment by dkelsey** [comment]
Thanks everyone!  We have more than one Kong setup in different environment, using different versions of Kong.  The exact error message I saw initially was from a newer version, 1.4.x.  and not what I was describing here, 0.14.1.  I've "inherited" this and have been focused on higher priority items.  This was very helpful regardless!
